lib/types/bcachefs_filesystem.nix

```diff
@@ -0,0 +1,394 @@
+      default =
+        let
+          # Generate a deterministic but random-looking UUID based on the filesystem name
+          # This avoids the need for impure access to nixpkgs at evaluation time
+          hash = builtins.hashString "sha256" "${config.name}";
+          hexChars = builtins.substring 0 32 hash;
+          p1 = builtins.substring 0 8 hexChars;
+          p2 = builtins.substring 8 4 hexChars;
+          p3 = builtins.substring 12 4 hexChars;
+          p4 = builtins.substring 16 4 hexChars;
+          p5 = builtins.substring 20 12 hexChars;
+        in
+        "${p1}-${p2}-${p3}-${p4}-${p5}";
+      defaultText = "generated deterministically based on filesystem name";
+      example = "809b3a2b-828a-4730-95e1-75b6343e415a";
+      description = ''
+        The UUID of the bcachefs filesystem.
+        If not provided, a deterministic UUID will be generated based on the filesystem name.
+      '';
+    };
+    passwordFile = lib.mkOption {
+      type = lib.types.nullOr diskoLib.optionTypes.absolute-pathname;
+      default = null;
+      description = ''
+        Path to the file containing the password for encryption.
+        Setting this option will automatically cause the `--encrypted` option to be passed to `bcachefs format` and cause the filesystem to have encryption enabled.
+      '';
+      example = "/tmp/disk.key";
+    };
+    subvolumes = lib.mkOption {
+      type = lib.types.attrsOf (
+        lib.types.submodule (
+          { config, ... }:
+          {
+            options = {
+              name = lib.mkOption {
+                type = lib.types.str;
+                default = config._module.args.name;
+                description = ''
+                  Path of the subvolume.
+                  Leading forward slashes are automatically removed.
+                '';
+                example = "subvolumes/home";
+              };
+              type = lib.mkOption {
+                type = lib.types.enum [ "bcachefs_subvolume" ];
+                default = "bcachefs_subvolume";
+                internal = true;
+                description = "Type";
+              };
+              mountOptions = lib.mkOption {
+                type = lib.types.listOf lib.types.str;
+                default = lib.naturalSort [
+                  "X-mount.mkdir"
+                  "X-mount.subdir=${lib.removePrefix "/" config.name}"
+                ];
+                description = ''
+                  Options to pass to mount.
+                  The "X-mount.mkdir" and "X-mount.subdir" options are always automatically added.
+                '';
+              };
+              mountpoint = lib.mkOption {
+                type = lib.types.nullOr diskoLib.optionTypes.absolute-pathname;
+                default = null;
+                description = ''
+                  Path to mount the subvolume to.
+                  DO NOT USE. Currently not working.
```

Comment:
So currently subvolumes are mounted relative to their parent or are they not mounted at all?

Reply:
Subvolumes are contained and exist within the filesystem they are created from. The `mount` command isn't called for them so they won't have a separate mountpoint, unless the `mountpoint` option is set, in which case the command will be ran, but mounting will fail at boot time in that case.

I've modified the description to state that that option is currently not working and should not be set though.